sessions dev

yevhenprudnik opened the pull request:
Improved device field sessions table; get all sessions,  remove specific session routes

yevhenprudnik committed:
sessions dev

yevhenprudnik committed:
test commit

evgeniidatsiuk commented on src/auth/auth.controller.ts:
```diff
@@ -51,42 +60,52 @@ export class AuthController {
   @Get('oauth/google')
   @UseGuards(AuthGuard('google'))
-  async googleAuth(@Req() req){}
+  googleAuth(){}
```
for what it?

evgeniidatsiuk commented on src/auth/auth.controller.ts:
```diff
@@ -51,42 +60,52 @@ export class AuthController {
 
+  @Get('sessions')
+  @UseGuards(TokenAuthGuard)
+  getSession(@Req() request){
```
getSessions

evgeniidatsiuk commented on src/auth/auth.controller.ts:
```diff
@@ -51,42 +60,52 @@ export class AuthController {
-  async refresh(@Req() request, @Res({ passthrough: true }) response){
+  async refresh(@Req() request, @Res({ passthrough: true }) response, @Headers('User-Agent') headers){
     const tokenFromUser = request.cookies.refreshToken;
+    const parsedUA = parser(headers);
```
why ua?
can we parse for other countries as well?

evgeniidatsiuk commented on src/auth/auth.controller.ts:
```diff
@@ -34,8 +42,9 @@ export class AuthController {
-    return this.sessionHandler.removeSession(request.user.id);
+  logOut(@Req() request, @Headers('User-Agent') headers, @Query('id') id): object{
+    const parsedUA = parser(headers);
+    return this.sessionHandler.removeSession(request.user.id, parsedUA, id);
```
you don't need parse headers here

evgeniidatsiuk commented on src/auth/auth.controller.ts:
```diff
@@ -1,31 +1,39 @@
-    const { refreshToken, ...rest } = await this.authService.logIn(userDto);
+  async logIn(@Body() userDto: LoginDto, @Res({ passthrough: true }) response, @Headers('User-Agent') headers){
+
+    const parsedUA = parser(headers);
```
it's not very readable, use another name for this, for example: `deviceInfo`, `systemInfo`, `device`, `system`

evgeniidatsiuk commented on src/auth/handlers/session.handler.ts:
```diff
@@ -13,16 +14,17 @@ export class SessionHandler {
       .createQueryBuilder("session")
       .innerJoinAndSelect("session.user", "user")
       .where("user.id = :id", { id: user.id })
+      .where(`session.device ::jsonb @> \'{"sessionId":${hashed}}\'`)
```
if use have id then use don't need use session id

yevhenprudnik replied:
I'm using session id to know if the user already have session on his devise. I need sessionId to identify the device, and if user doesn't have that session I'll create another for specific device. Only user Id isn't enough.
Is it ok?

evgeniidatsiuk commented on src/interfaces/device.interface.ts:
```diff
@@ -0,0 +1,6 @@
+export interface deviceInterface {
```
first letter must be upper case

evgeniidatsiuk commented on src/entities/session.entity.ts:
```diff
@@ -16,10 +17,9 @@ export class Session {
+    type: "jsonb",
   })
-  device: string;
+  device: deviceInterface;
```
don't use interface, use class for that in session.entity file

yevhenprudnik replied:
Why is it better to use class?

evgeniidatsiuk replied:
Because use can add decorators )

yevhenprudnik replied:
I've added separate decorator for the endpoint , is it ok?

evgeniidatsiuk commented on src/auth/handlers/session.handler.ts:
```diff
@@ -75,29 +80,29 @@ export class SessionHandler {
       .innerJoinAndSelect("session.user", "user")
-      .where("user.id = :id", { id: userId })
+      .where("user.id = :id", { id: userId})
+      .where(`session.device ::jsonb @> \'{"sessionId":${sessionId}}\'`)
```
if use have id then use don't need use session id

evgeniidatsiuk commented on src/auth/auth.service.ts:
```diff
@@ -22,7 +22,7 @@ export class AuthService {
    * @returns object with user id and accessToken
    */
-  async register(userDto: RegisterDto){
+  async register(userDto: RegisterDto, parsedUA){
```
`userDto: RegisterDto` - **bad**
`definition: DeepPartial<User>` - **good**

evgeniidatsiuk replied:
in service we don't import dto class

evgeniidatsiuk commented on src/auth/oauth/oauth.service.ts:
```diff
@@ -51,11 +54,18 @@ export class OauthService {
+    }
+    const tokens = await this.sessionHandler.createSession(request.user, parsedUA);
+
+    return tokens;
```
return this.sessionHandler.createSession(request.user, parsedUA);

evgeniidatsiuk commented on src/auth/auth.controller.ts:
```diff
@@ -51,42 +60,52 @@ export class AuthController {
-  async googleAuthRedirect(@Req() request, @Res({ passthrough: true }) response): Promise<object>{
-
-    const { refreshToken, ...rest } = await this.sessionHandler.getSession(request.user.id);
+  async googleAuthRedirect(@Req() request, @Res({ passthrough: true }) response, @Headers('User-Agent') headers): Promise<object>{
```
don't use object

evgeniidatsiuk commented on src/auth/auth.controller.ts:
```diff
@@ -51,42 +60,52 @@ export class AuthController {
+  async googleAuthRedirect(@Req() request, @Res({ passthrough: true }) response, @Headers('User-Agent') headers): Promise<object>{
 
-    response.cookie( 'refreshToken', refreshToken );
+    const parsedUA = parser(headers);
```
make decorator for this parser, don't copy paste the same code for each function)


`googleAuthRedirect(@Req() request, @SystemInfo() systemInfo) Promise<Session>{}`

yevhenprudnik committed:
minor changes

yevhenprudnik committed:
dep

evgeniidatsiuk commented on src/auth/auth.controller.ts:
```diff
@@ -51,42 +57,45 @@ export class AuthController {
   @Get('oauth/google')
   @UseGuards(AuthGuard('google'))
-  async googleAuth(@Req() req){}
+  googleAuth(){}
```
for what this function?

yevhenprudnik replied:
this function is required for AuthGuard to be able to redirect user in case of successful authentication

yevhenprudnik committed:
using hashed user device info as Primary Column in session

yevhenprudnik committed:
added validation on oauth2 session

yevhenprudnik committed:
added validation on oauth2 session(access token expiration)

yevhenprudnik commented:
hello, @evgeniidatsiuk )
- Added user validation on generateSession for oauth2(via access token provided by google/facebook)
- Added username to hashed session id(to prevent id duplication for users with identical devices). 
- Sorted oauth URLs

yevhenprudnik committed:
dependency clean, adding bidirectional relations

yevhenprudnik committed:
redo oauth

evgeniidatsiuk commented on src/auth/handlers/oauth.handelr.ts:
```diff
@@ -0,0 +1,98 @@
+      
+      const user = await this.usersRepository.save(newUser);
+      
+      await this.mailerService.sendActivationEmail(profile.email, emailConfirmationLink);
```
don't need to verify social accounts ))

evgeniidatsiuk commented on src/auth/auth.service.ts:
```diff
@@ -18,12 +16,12 @@ export class AuthService {
    */
-  async register(userDto: RegisterDto){
-    const { username, email, password, firstName, lastName } = userDto;
+  async register(userDefinition: DeepPartial<User>, systemInfo){
```
can be just definition

evgeniidatsiuk commented on src/auth/auth.service.ts:
```diff
@@ -86,13 +84,13 @@ export class AuthService {
-    const tokens = await this.sessionHandler.createSession(user);
+    const tokens = await this.sessionHandler.createSession(user, systemInfo);
 
     return { userId: user.id, ...tokens };
```
`return this.sessionHandler.createSession(user, systemInfo);`

evgeniidatsiuk commented on src/auth/auth.service.ts:
```diff
@@ -77,7 +75,7 @@ export class AuthService {
-    const tokens = await this.sessionHandler.createSession(candidate);
+    const tokens = await this.sessionHandler.createSession(candidate, systemInfo);
 
     return { userId: candidate.id, ...tokens };
```
`return this.sessionHandler.createSession(user, systemInfo);`

evgeniidatsiuk commented on src/auth/auth.service.ts:
```diff
@@ -52,18 +50,18 @@ export class AuthService {
    */
-  async logIn(userDto: LoginDto){
-    const { email, password } = userDto;
+  async logIn(userDefinition: DeepPartial<User>, systemInfo){
```
add `Promise<Session>`

evgeniidatsiuk commented on src/decorators/user-agent.decorator.ts:
```diff
@@ -0,0 +1,10 @@
+import { createParamDecorator, ExecutionContext } from '@nestjs/common';
```
rename to `system-info.ts`

evgeniidatsiuk commented on src/entities/session.entity.ts:
```diff
@@ -16,10 +20,9 @@ export class Session {
+    type: "jsonb",
   })
-  device: string;
+  device: object;
```
don't use object.

yevhenprudnik committed:
fixed todos

evgeniidatsiuk commented on src/auth/auth.controller.ts:
```diff
@@ -51,42 +56,41 @@ export class AuthController {
 
-    response.cookie( 'refreshToken', refreshToken );
+    const tokens = await this.oauthHandler.oauthHandler('facebook', AuthHeaders, systemInfo);
+    response.cookie( 'refreshToken', tokens.refreshToken );
```
set also accessToken to cookies

evgeniidatsiuk commented on src/auth/auth.controller.ts:
```diff
@@ -1,41 +1,46 @@
+    response.cookie('refreshToken', refreshToken);
     
-    return rest;
+    return accessToken;
```
why only access token? how a mobile app can refresh session? they don't have cookies

evgeniidatsiuk commented on src/interfaces/user.oauth.interface.ts:
```diff
@@ -1,4 +1,5 @@
 export interface userOauth {
```
Upper case

evgeniidatsiuk commented on src/auth/handlers/oauth.handelr.ts:
```diff
@@ -0,0 +1,93 @@
+@Injectable()
+export class OauthHandler {
+  constructor(
+    @InjectRepository(User) 
```
don't inject repository, use can use auth service

yevhenprudnik replied:
should I move authUser function to auth service?

evgeniidatsiuk commented on src/auth/handlers/oauth.handelr.ts:
```diff
@@ -0,0 +1,93 @@
+    .createQueryBuilder('user')
+    .where(`user.oauth ::jsonb @> \'{"id":"${profile.id}"}\'`)
+    .getOne();
+  if(user){
```
`if (user) {
    return this.sessionHandler.createSession(user, systemInfo)
}`

remove else condition, just write code after return without `else {}`

evgeniidatsiuk commented on src/auth/handlers/oauth.handelr.ts:
```diff
@@ -0,0 +1,93 @@
+  } else {
+      const username = profile.name.toLocaleLowerCase().replace(/\s/g, '');
+      const candidateByUsername = await this.usersRepository.findOneBy({ username });
+      if (candidateByUsername){
+        throw new BadRequestException(`Looks like username ${username} is already taken, please try common method of registration and come up with different username`);
+      }
+      const candidateByEmail = await this.usersRepository.findOneBy({ email : profile.email });
+      if (candidateByEmail){
+        throw new BadRequestException(`Looks like username with email ${profile.email} have already been registered via another authentication method. Please use your initial type of authentication`);
```
maybe is better to use oauth ID instead of username?

yevhenprudnik replied:
I'm using auth ID at first check. What if user with such username already exist? Shouldn't username be unique?

evgeniidatsiuk commented on src/auth/handlers/oauth.handelr.ts:
```diff
@@ -0,0 +1,93 @@
+import { User } from '../../entities/user.entity';
+import { EmailHandler } from './mail.handler';
+import { SessionHandler } from './session.handler';
+@Injectable()
```
1. add space after import )))
2. rename file to `oauth.handler.ts`

evgeniidatsiuk commented on src/auth/auth.controller.ts:
```diff
@@ -1,41 +1,46 @@
 
   @Post('register')
-  async register(@Body() userDto: RegisterDto, @Res({ passthrough: true }) response){
+  async register( @Body() userDto: RegisterDto, @SystemInfo() systemInfo, @Res({ passthrough: true }) response){
```
`userDto: UserDto`
`registerDto: RegisterDto`
make sure that variable has the same name as in class, it's more redable

**can be also just** `input: UserDto` - it's standart

yevhenprudnik committed:
tokens response/cookies

yevhenprudnik committed:
register refresh token response

evgeniidatsiuk commented on src/auth/auth.service.ts:
```diff
@@ -1,29 +1,28 @@
-    private readonly usersRepository: Repository<User>,
-    private sessionHandler: SessionHandler,
-    private mailerService: EmailHandler
+    public usersRepository: Repository<User>,
```
make private readonly

evgeniidatsiuk commented on src/auth/auth.service.ts:
```diff
@@ -1,29 +1,28 @@
-    private sessionHandler: SessionHandler,
-    private mailerService: EmailHandler
+    public usersRepository: Repository<User>,
+    public sessionHandler: SessionHandler,
```
private readonly

evgeniidatsiuk commented on src/auth/auth.service.ts:
```diff
@@ -1,29 +1,28 @@
-    private mailerService: EmailHandler
+    public usersRepository: Repository<User>,
+    public sessionHandler: SessionHandler,
+    public mailerService: EmailHandler
```
private readonly

evgeniidatsiuk commented on src/auth/guards/token.auth.guard.ts:
```diff
@@ -9,13 +9,15 @@ export class TokenAuthGuard implements CanActivate{
+      
+      if(!accessToken){
+        const authHeaders = authHeader.split(" "); 
+        if (authHeaders[0] !== "Bearer" || !authHeaders[1]) {
```
`if authHeader.startsWith('Bearer')` ...

yevhenprudnik committed:
private services+startsWith

evgeniidatsiuk commented on src/auth/auth.service.ts:
```diff
@@ -126,4 +121,89 @@ export class AuthService {
+   * @param  {any} profile
+   * user profile from the provider
+   */
+  async oauthUserHandler(profile: any, provider: string, systemInfo): Promise<Session>{
```
don't use `any`. write interface for that

evgeniidatsiuk commented on src/auth/auth.service.ts:
```diff
@@ -126,4 +121,89 @@ export class AuthService {
+    }
+    const username = await this.generateUsername(profile.name);
+    if (!username){
+      throw new BadRequestException(`Please try common method of registration and come up with unique username`);
```
are you sure that  it is okay flow for social login?

evgeniidatsiuk commented on src/entities/user.entity.ts:
```diff
@@ -50,14 +53,17 @@ export class User {
     default: {}
   })
-  oauth: userOauth;
+  oauth: UserOauth;
```
don't use interface

evgeniidatsiuk commented on src/entities/user.entity.ts:
```diff
@@ -50,14 +53,17 @@ export class User {
     default: {}
   })
-  additionalInfo: userAdditionalInfo;
+  additionalInfo: UserAdditionalInfo;
```
don't use interface

evgeniidatsiuk commented on src/helpers/username.functions.ts:
```diff
@@ -0,0 +1,10 @@
+export const usernameFunctions = [
+  profileName => profileName.replace(/\s/g, ''),
+  profileName => profileName.replace(/\s/g, '_'),
+  profileName => profileName.replace(/\s/g, '.'),
+  profileName => profileName.replace(/\s/g, '-'),
+  profileName => profileName.toLocaleLowerCase().replace(/\s/g, ''),
+  profileName => profileName.toLocaleLowerCase().replace(/\s/g, '_'),
+  profileName => profileName.toLocaleLowerCase().replace(/\s/g, '.'),
+  profileName => profileName.toLocaleLowerCase().replace(/\s/g, '-'),
+]
```
can you write tests for that function?

evgeniidatsiuk commented on src/auth/auth.service.ts:
```diff
@@ -126,4 +121,89 @@ export class AuthService {
+  async generateUsername(profileName: string){
+    for(const fn of usernameFunctions){
+      const username = fn(profileName);
+      const candidateByUsername = await this.usersRepository.findOneBy({ username });
```
use `for await `
use `count` instead of `findOne`

yevhenprudnik committed:
changing oauth flow

yevhenprudnik committed:
deleting username functions file(unused)

evgeniidatsiuk commented on src/schemas/additionalInfo.schema.ts:
```diff
@@ -1,4 +1,4 @@
-export interface userAdditionalInfo {
+export class UserAdditionalInfo {
```
move it to entity class

evgeniidatsiuk commented on src/schemas/device.schema.ts:
```diff
@@ -0,0 +1,5 @@
+export class DeviceSchema {
```
move it to entity class

evgeniidatsiuk commented on src/schemas/oauthInfo.schema.ts:
```diff
@@ -0,0 +1,5 @@
+export class UserOauth {
```
move it to entity class

evgeniidatsiuk commented on src/interfaces/oauth.profile.interface.ts:
```diff
@@ -0,0 +1,10 @@
+  family_name?: string,
+  first_name?: string,
+  last_name?: string,
+  picture?
```
picture: { data: { url: string } }

evgeniidatsiuk commented on src/dtos/username.dto.ts:
```diff
@@ -0,0 +1,6 @@
+import { IsNotEmpty } from "class-validator";
+
+export class UsernameDto {
```
username dto?))

evgeniidatsiuk commented on src/auth/auth.service.ts:
```diff
@@ -126,4 +121,82 @@ export class AuthService {
+   * user profile from the provider
+   */
+  async oauthUserHandler(profile: OauthProfileInterface, provider: string, systemInfo): Promise<Session>{
+    const userFromDb = await this.usersRepository
```
could be written as `user`?

evgeniidatsiuk commented on src/auth/auth.service.ts:
```diff
@@ -126,4 +121,82 @@ export class AuthService {
+    if(userFromDb){
+      return this.sessionHandler.createSession(userFromDb, systemInfo)
+    } 
+    const candidateByEmail = await this.usersRepository.findOneBy({ email : profile.email });
```
use count instead of find

evgeniidatsiuk commented on src/interfaces/oauth.profile.interface.ts:
```diff
@@ -0,0 +1,10 @@
+export interface OauthProfileInterface {
```
FacebookUser

evgeniidatsiuk replied:
Also add interface for GoogleUser

yevhenprudnik committed:
schemas -> entity Google/Facebook create user decomposition

yevhenprudnik committed:
adding space

evgeniidatsiuk commented on src/auth/handlers/session.handler.ts:
```diff
@@ -2,22 +2,28 @@ import { Injectable, UnauthorizedException } from '@nestjs/common';
+import hashFunction from '../../helpers/string.hash.generator'
 
 @Injectable()
 export class SessionHandler {
```
It's session service. Rename to SessionService.

evgeniidatsiuk commented on src/auth/handlers/session.handler.ts:
```diff
@@ -2,22 +2,28 @@ import { Injectable, UnauthorizedException } from '@nestjs/common';
     @InjectRepository(Session) 
     private readonly sessionRepository: Repository<Session>,
+    @InjectRepository(User) 
+    private readonly userRepository: Repository<User>,
```
Inject Auth service instead of injection the repository

yevhenprudnik committed:
changing file/class names

yevhenprudnik committed:
comments

evgeniidatsiuk commented on src/auth/session.service.ts:
```diff
@@ -26,78 +35,75 @@ export class SessionHandler {
+   * @param  {} sessionId
+   * (optional) id of a session to remove
+   */
+  async removeSession(user: User, systemInfo, sessionId?){
```
session id?

evgeniidatsiuk commented on src/auth/user.controller.ts:
```diff
@@ -0,0 +1,129 @@
+  @Get('sessions')
+  @UseGuards(TokenAuthGuard)
+  getSessions(@Req() request){
+    return this.userService.getAllSession(request.user);
```
getAllSession**s**

yevhenprudnik committed:
adding types
